library/entropy_poll.c

```diff
@@ -85,11 +85,10 @@ int mbedtls_platform_entropy_poll( void *data, unsigned char *output, size_t len
 #else /* _WIN32 && !EFIX64 && !EFI32 */
 
 /*
- * Test for Linux getrandom() support.
- * Since there is no wrapper in the libc yet, use the generic syscall wrapper
- * available in GNU libc and compatible libc's (eg uClibc).
+ * Test for Linux getrandom() support. Since many libc implementations on
+ * Linux don't have a wrapper for getrandom(), issue the system call directly.
  */
-#if ((defined(__linux__) && defined(__GLIBC__)) || defined(__midipix__))
+#if defined(__linux__) || defined(__midipix__)
```

Comment:
Hum, I [objected to a similar change](https://github.com/ARMmbed/mbedtls/pull/3642#pullrequestreview-491353564) in #3642 and I'm going to have the same objection here: what guarantee do we have that it's OK to `#include <sys/syscall.h>` and use `syscall()` as soon as `__linux__` is defined, regardless of what libc is used? I agree with the goal of using the `getrandom` syscall with more libcs, but I don't think we can just be optimistic about it when the consequence if we're wrong is a broken build.

Unless there's something I missed that says this header and the `syscall()` function are always present with any libc as soon as the target is Linux?

Reply:
I verified dietlibc (works but needs `-lcompat`) and musl, and it looks ok with bionic and uclibc by source inspection. What other libc should I check?

I thought of changing `__GLIBC__` to a list of symbols but I can't find a way to detect musl.

Reply:
> What other libc should I check?

I don't know but that's kind of my point. What if somebody uses a libc we've never heard of? What if a new libc is created / becomes popular that doesn't have `sys/syscall.h` or `syscall()`? With this patch, people using such a libc would just get a build failure. Currently they get a working build (just using `/dev/urandom` instead of `getrandom()`). IMO this is a regression. Admittedly, it's a regression that only happens in a rare to hypothetical case, but I'm not entirely comfortable discarding that case entirely either.

It's annoying that there's no easy way to detect musl. It leaves us with a difficult choice - either have a list of libraries known to have `syscall`, but then miss out musl, or just assume that `syscall` is available, and then have a build failure if and when it isn't available...

Reply:
Well, actually there's another possibility, though I don't like it very much: we could have something like `MBEDTLS_HAVE_SYSCALL` (similarly to `MBEDTLS_HAVE_ASM` etc) in `config.h`, which would only have an effect on Linux, and would be set by default. Or perhaps rather a negative option, for compatibility with existing standalone configs. I'm really not liking that idea, I'm just mentioning it for the sake of completeness.

Reply:
Yes. This is why other projects have a configure script.